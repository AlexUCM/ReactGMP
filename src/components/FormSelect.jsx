import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { useClickOutside } from '../hooks/useClickOutside';
import styled from 'styled-components';

const Container = styled.div`
  position: relative;
  margin: 8px 0 0 0;
`;

const Select = styled.div`
  display: flex;
  justify-content: space-between;
  padding: 0 16px;
  background-color: #232323c9;
  border-radius: 5px;
  font-size: 24px;
  line-height: 48px;
  cursor: pointer;
`;

const Value = styled.p`
  color: #fff;
  text-transform: capitalize;
  white-space: nowrap;
  overflow-x: hidden;
  text-overflow: ellipsis;
  user-select: none;

  ${({ value }) =>
    !value &&
    `
    color: #a9a9a9c0;
  `}
`;

const Arrow = styled.div`
  color: #f65261;
  font-size: 16px;

  ${({ isOpen }) =>
    isOpen &&
    `
    transform: rotate(180deg);
  `}
`;

const Dropdown = styled.div`
  position: absolute;
  display: none;
  left: 0;
  top: 50px;
  width: 100%;
  max-height: 210px;
  overflow-y: scroll;
  border-radius: 5px;
  background-color: #232323;
  z-index: 50;

  ${({ isOpen }) =>
    isOpen &&
    `
    display: block;
  `}
`;

const InputWrapper = styled.div`
  display: flex;
  align-items: baseline;
  font-size: 22px;
  line-height: 42px;
  color: #fff;
`;

const Input = styled.input`
  display: none;

  &:checked + label {
    color: #fff;
  }

  &:checked + label::before {
    border-color: #f65261;
    background-color: #f65261;
    background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 8 8'%3e%3cpath fill='%23fff' d='M6.564.75l-3.59 3.612-1.538-1.55L0 4.26 2.974 7.25 8 2.193z'/%3e%3c/svg%3e");
  }
`;

const Label = styled.label`
  display: inline-flex;
  align-items: center;
  user-select: none;
  color: #a9a9a9c0;

  &:before {
    content: '';
    width: 14px;
    height: 14px;
    border: 1px solid #adb5bd;
    border-radius: 4px;
    margin: 0 8px 0 12px;
    background-repeat: no-repeat;
    background-position: center center;
    background-size: 75% 75%;
    background-color: #fff;
  }

  & span {
    text-transform: capitalize;
  }
`;

const genres = [
  'action',
  'adventure',
  'animation',
  'comedy',
  'crime',
  'documentary',
  'drama',
  'family',
  'fantasy',
  'history',
  'horror',
  'music',
  'mystery',
  'romance',
  'science fiction',
  'thriller',
  'war',
  'western',
];

export const FormSelect = ({ name, value, onChange }) => {
  const ref = useRef();
  const [isOpen, setIsOpen] = useState(false);
  useClickOutside(ref, () => setIsOpen(false));

  return (
    <>
      <span>GENRE</span>
      <Container ref={ref}>
        <Select onClick={() => setIsOpen((state) => !state)}>
          <Value value={value}>{value || 'Select genre'}</Value>
          <Arrow isOpen={isOpen}>&#9660;</Arrow>
        </Select>
        <Dropdown data-testid='dropdown' isOpen={isOpen}>
          {genres.map((genre) => (
            <InputWrapper key={genre}>
              <Input
                id={genre}
                name={name}
                type='checkbox'
                checked={value.split(', ').includes(genre)}
                value={genre}
                onChange={onChange}
              />
              <Label htmlFor={genre}>
                <span>{genre}</span>
              </Label>
            </InputWrapper>
          ))}
        </Dropdown>
      </Container>
    </>
  );
};

FormSelect.propTypes = {
  value: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};
