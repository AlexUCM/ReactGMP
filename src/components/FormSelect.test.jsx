import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from '@testing-library/react';
import { FormSelect } from './FormSelect';

describe('FormSelect', () => {
  const getComponent = ({
    value = '',
    name = 'genres',
    onChange = jest.fn(),
  }) => render(<FormSelect value={value} name={name} onChange={onChange} />);
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot();
  });
  it('Should open dropdown', () => {
    const { getByText, getByTestId } = getComponent({});
    const select = getByText('Select genre');
    const dropdown = getByTestId('dropdown');
    fireEvent.click(select);
    expect(dropdown).toHaveStyle('display: block');
  });
  it('Should close dropdown', () => {
    const { getByText, getByTestId } = getComponent({});
    const select = getByText('Select genre');
    const dropdown = getByTestId('dropdown');
    fireEvent.click(select);
    fireEvent.click(document);
    expect(dropdown).toHaveStyle('display: none');
  });
  it('Should call onChange', () => {
    const onChange = jest.fn();
    const { getByText, getByLabelText } = getComponent({ onChange });
    const select = getByText('Select genre');
    const checkbox = getByLabelText('comedy');
    fireEvent.click(select);
    fireEvent.click(checkbox);
    expect(onChange).toBeCalled();
  });
});
