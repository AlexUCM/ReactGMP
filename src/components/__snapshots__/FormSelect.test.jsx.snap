// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect Should renders correctly 1`] = `
<DocumentFragment>
  <span>
    GENRE
  </span>
  <div
    class="sc-bdfBwQ gXNUXp"
  >
    <div
      class="sc-gsTCUz ekbeGm"
    >
      <p
        class="sc-dlfnbm bbSCXp"
        value=""
      >
        Select genre
      </p>
      <div
        class="sc-hKgILt cHpNUq"
      >
        â–¼
      </div>
    </div>
    <div
      class="sc-eCssSg zNtVm"
      data-testid="dropdown"
    >
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="action"
          name="genres"
          type="checkbox"
          value="action"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="action"
        >
          <span>
            action
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="adventure"
          name="genres"
          type="checkbox"
          value="adventure"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="adventure"
        >
          <span>
            adventure
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="animation"
          name="genres"
          type="checkbox"
          value="animation"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="animation"
        >
          <span>
            animation
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="comedy"
          name="genres"
          type="checkbox"
          value="comedy"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="comedy"
        >
          <span>
            comedy
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="crime"
          name="genres"
          type="checkbox"
          value="crime"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="crime"
        >
          <span>
            crime
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="documentary"
          name="genres"
          type="checkbox"
          value="documentary"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="documentary"
        >
          <span>
            documentary
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="drama"
          name="genres"
          type="checkbox"
          value="drama"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="drama"
        >
          <span>
            drama
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="family"
          name="genres"
          type="checkbox"
          value="family"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="family"
        >
          <span>
            family
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="fantasy"
          name="genres"
          type="checkbox"
          value="fantasy"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="fantasy"
        >
          <span>
            fantasy
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="history"
          name="genres"
          type="checkbox"
          value="history"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="history"
        >
          <span>
            history
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="horror"
          name="genres"
          type="checkbox"
          value="horror"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="horror"
        >
          <span>
            horror
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="music"
          name="genres"
          type="checkbox"
          value="music"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="music"
        >
          <span>
            music
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="mystery"
          name="genres"
          type="checkbox"
          value="mystery"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="mystery"
        >
          <span>
            mystery
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="romance"
          name="genres"
          type="checkbox"
          value="romance"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="romance"
        >
          <span>
            romance
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="science fiction"
          name="genres"
          type="checkbox"
          value="science fiction"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="science fiction"
        >
          <span>
            science fiction
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="thriller"
          name="genres"
          type="checkbox"
          value="thriller"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="thriller"
        >
          <span>
            thriller
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="war"
          name="genres"
          type="checkbox"
          value="war"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="war"
        >
          <span>
            war
          </span>
        </label>
      </div>
      <div
        class="sc-jSgupP kDLaHj"
      >
        <input
          class="sc-gKsewC lamki"
          id="western"
          name="genres"
          type="checkbox"
          value="western"
        />
        <label
          class="sc-iBPRYJ gDGHSk"
          for="western"
        >
          <span>
            western
          </span>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
